lib/new_relic/agent/serverless_handler.rb

```diff
@@ -146,15 +239,142 @@ def use_named_pipe?
+        return unless category == :web
+
+        if @http_uri
+          add_agent_attribute('uri.host', @http_uri.host)
+          add_agent_attribute('uri.port', @http_uri.port)
+          if NewRelic::Agent.instance.attribute_filter.allows_key?('http.url', AttributeFilter::DST_SPAN_EVENTS)
+            add_agent_attribute('http.url', @http_uri.to_s)
+          end
+        end
+
+        if @http_method
+          add_agent_attribute('http.method', @http_method)
+          add_agent_attribute('http.request.method', @http_method)
+        end
+      end
+
+      def api_gateway_event?
+        return false unless @event
+
+        # '1.0' for API Gateway V1, '2.0' for API Gateway V2
+        return true if @event.fetch('version', '').start_with?(DIGIT)
+
+        return false unless headers_from_event
+
+        # API Gateway V1 - look for toplevel 'path' and 'httpMethod' keys if a version is unset
+        return true if @event.fetch('path', nil) && @event.fetch('httpMethod', nil)
+
+        # API Gateway V2 - look for 'requestContext/http' inner nested 'path' and 'method' keys if a version is unset
+        return true if @event.dig('requestContext', 'http', 'path') && @event.dig('requestContext', 'http', 'method')
+
+        false
+      end
+
+      def add_event_source_attributes
+        arn = event_source_arn
+        add_agent_attribute('aws.lambda.eventSource.arn', arn) if arn
+
+        info = event_source_event_info
+        return unless info
+
+        add_agent_attribute('aws.lambda.eventSource.eventType', info['name'])
+
+        info['attributes'].each do |name, elements|
+          next if elements.empty?
+
+          size = false
+          if elements.last.eql?('#size')
+            elements = elements.dup
+            elements.pop
+            size = true
+          end
+          value = @event.dig(*elements)
+          value = value.size if size
+          next unless value
+
+          add_agent_attribute(name, value)
+        end
+      end
+
+      def event_source_arn
+        return unless @event
+
+        # SQS/Kinesis Stream/DynamoDB/CodeCommit/S3/SNS
+        return event_source_arn_for_records if @event.fetch('Records', nil)
+
+        # Kinesis Firehose
+        ds_arn = @event.fetch('deliveryStreamArn', nil) if @event.fetch('records', nil)
+        return ds_arn if ds_arn
+
+        # ELB
+        elb_arn = @event.dig('requestContext', 'elb', 'targetGroupArn')
+        return elb_arn if elb_arn
+
+        # (other)
+        es_arn = @event.dig('resources', 0)
+        return es_arn if es_arn
+
+        NewRelic::Agent.logger.debug 'Unable to determine an event source arn'
+
+        nil
+      end
+
+      def event_source_event_info
+        return unless @event
+
+        # if every required key for a source is found, consider that source
+        # to be a match
+        EVENT_SOURCES.each do |_type, info|
+          return info unless info['required_keys'].detect { |r| @event.dig(*r).nil? }
+        end
```

Comment:
What about using `each_value` here instead? [Ruby 2.4 should support it](https://ruby-doc.org/core-2.4.10/Hash.html#method-i-each_value).

 ```ruby
 EVENT_SOURCES.each_value do |info|
          return info unless info['required_keys'].detect { |r| @event.dig(*r).nil? }
        end
 ```
 (Rubocop gave me this rec recently on an OTel repo: https://docs.rubocop.org/rubocop/cops_style.html#stylehasheachmethods; benchmarks show a same-ish result between the two options)

Reply:
Good call. In an earlier iteration of this code, I noticed the `_type` silliness and tried a refactor. I might have used `.values.each` instead of your `each_value`, but it was something similar. For some reason, doing so broke the functionality. However, this was in an earlier version of the code than what made it to the PR and it was doing more than this one-liner does. Using `each_value` here with the one-liner has been tested to work great. Thanks! Added with [af52a76](https://github.com/newrelic/newrelic-ruby-agent/pull/2811/commits/af52a76ae743ea8629c6c9138d6077b753e5ab50).